File: elixir/square-root/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand) do
    1..radicand
    |> Stream.filter(&(&1 ** 2 == radicand))
    |> Enum.take(1)
    |> hd()
  end
end

File: c/square-root/square_root.c
#include "square_root.h"

static const double epsilon = 1e-4;

double converge(const double x, const double n);

double converge(const double x, const double n) {
  const double f = (x * x) - n;
  const double f_prime = 2 * x;
  return f / f_prime;
}

double square_root(double n) {
  double root = n;
  double c;
  do {
    c = converge(root, n);
    root -= c;
  } while (c > epsilon);
  return root;
}
